### PR TITLE
Added DiagnosticsSuppressor for NullReference

### LINKED ISSUE #157: Suppress nullability warnings after Is.Not.Null constraint
As discussed here https://github.com/nunit/nunit/issues/3376#issuecomment-535879642, it would be great to suppress nullability warnings after Is.Not.Null asserts, e.g.
```csharp
string? str = GetNullable();
Assert.That(str, Is.Not.Null);
Assert.That(str.Length, Is.Not.Zero);  // <----- should be no warning here
```

Unfortunately, we cannot impact roslyn's nullability analysis, only suppress analyzers (see https://github.com/dotnet/roslyn/issues/32042#issuecomment-540524650), which means that at least very basic logic should be re-invented.

Cases that I think should be covered:
`Assert.That(obj?.PropA?.PropB, Is.Not.Null)`
`Assert.That(obj, Is.Not.Null & Is.Not.Empty)`
`Assert.That(obj, Is.Not.Null.And.Not.Empty)`

```csharp
Assert.Multiple(() =>
{
	string? str = GetNullable();
	Assert.That(str, Is.Not.Null);
	Assert.That(str.Length, Is.Not.Zero);  // <----- warning should remain here
});
```

### PR DESCRIPTION
**Pull request by manfred-brands**
Fixes #157 

Preliminary version:
* Detects: Assert.NotNull, Assert.IsNotNull and Assert.That(..., Is.Not.Null)
* Recognizes: Assert.Multiple
* Recognizes: Assignments

The current version is certainly not catching all cases.
I welcome suggestion for other test cases.

### CODE REVIEW

**Comment by manfred-brands** [comment]
Added recognition that `Assert.That(a?.B?.C, Is.Not.Null)` means that all three of `a`, `a.B` and `a.B.C` are not null.
Added recognition that assigning to `a` or `b` will invalidate this.

Cleaned up the code a bit.
Added more unit test cases.

**Comment by manfred-brands** [comment]
Our .nuspec file was hardcoded to take the netstandard1.6 dll which does not support the DiagnosticSuppressor.
Roslyn analyzers cannot be multi-targeted. The [ticket](https://github.com/dotnet/roslyn-sdk/issues/107) has been open for almost 3 years. Roslyn itself has the 2.x range and the 3.x range. We could issue two sets of our package as well or drop support for VS2017.

For now I added an option to the build scripts to specify which dll to add to the nuget package: `-targetFramework netstandard2.0`
Depending the decision from @mikkelbu we will need to update the script to create two different nuget packages: 0.x range for VS2017 and 1.x range for VS2019 (or we could make the leap and use the same range as roslyn: 2.x and 3.x)

**Comment by manfred-brands** [comment]
After using it in one of our NUnit projects which needed converting to Nullability, there were more CS errors that needs suppressing. Added a few more use cases.
Otherwise it seems to work.

**Comment by manfred-brands** [comment]
The last commit adds a suppressor for nullable fields not initialized. The rule is suppressed if the field is initialized in the (OneTime)SetUp method. This helps for test setting class instances in the Setup of the test.

**Comment by mikkelbu** [comment]
First and foremost, I very much appreciate this contribution, and I hope to have time to review it in detail the one of the following days (so far I've only skimmed the files).

I still think we should support VS2017, and I like your approach of being able to specify the targetFramework. I'm unsure of whether the best approach would be to have two separate feeds or one feed with two different versions numbers, but I'll give it some more thought.

File: src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs
Comment (mikkelbu) [inline src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs:69]: Perhaps we should also handle when the left-side is prefixed with `this`, e.g. `this.someVariable`?

File: src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs
Comment (mikkelbu) [inline src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs:36]: Probably my lack of understanding of `DiagnosticSuppressor`, but can we do something more strongly typed than a string for the field? (same comment for the other suppressor)

**Review by mikkelbu — COMMENTED** [review]
So far it looks good. I've tested the code a little and reviewed all but the tests (and `DiagnosticsSuppressorAnalyzer.cs`) - I'll see if I've time for the last part tomorrow night.

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DiagnosticsSuppressorAnalyzer.cs
Comment (mikkelbu) [inline src/nunit.analyzers.tests/DiagnosticSuppressors/DiagnosticsSuppressorAnalyzer.cs:106]:
```suggestion
```

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DiagnosticsSuppressorAnalyzer.cs
Comment (mikkelbu) [inline src/nunit.analyzers.tests/DiagnosticSuppressors/DiagnosticsSuppressorAnalyzer.cs:86]: I could not understand why we didn't need to check `programmaticSuppressionInfoProperty` for null, as my VS says that the type is `PropertyInfo?` (which is strange as the VS also states that `GetProperty` returns a `PropertyInfo`), and in the next line it says `'programmaticSuppressionInfoProperty' is not null here.`.

Is this because in ".NET Standard 2.1" the return type of `GetProperty` is `PropertyInfo`, so the compiler assumes that this cannot be null, but since I've also have ".NET 5.0" installed (which has `PropertyInfo?` as the return type), then VS shows me this information?

**Review by mikkelbu — APPROVED** [review]
Great work @manfred-brands 🥇 . I've only left a few comments and questions.

Regarding releasing this then I'm currently leaning towards your proposal of having two versions in the same feed (and use the same range as Roslyn - even though these version numbers are unknown for most people). But I think we should do this in a subsequent issue/PR to make the change more visible (with respect to release notes etc.).

I'm most happy to merge this when you are 👍

File: src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs
Comment (manfred-brands) [reply src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs:69]: I added two test cases, but the one advantage of text based comparison is that the code already handles `this.member` as the left part of the assignment.
The downside it, it doesn't handle inconsistent use of the `this` prefix, e.g. using it in the assignment, but not in references. But that would already be flagged by the compiler (IDE0009).

File: src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs
Comment (manfred-brands) [reply src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs:36]: CS8618 is raised for nulable fields, so we know that the SyntaxNode is a field. We don't actually care about the type.
But in this case, I do need to detect `this.field` as a separate assignment case: Code updated.

Comment (manfred-brands) [reply src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs:36]: For the actual null reference suppressor, we get the error reported on the expression. Again we don't care about the type. I check if the same expression that raises the null reference error is also used in the `Assert` call.
It fails when mixing expressions like in:
```
A? a = GetA(create);
Assert.That(a?.B?.Text, Is.Not.Null);
B b = a.B;
Assert.That(b.Text.Length, Is.GreaterThan(0));
```
Even though `b.Text` was tested as `a?.B?.Text` neither string comparison nor symbol analysis would help. Only if we could analyze the flow of assignments. The example is contrived, a more appropriate example would be:
```
A? a = GetA(create);
string? t = a?.B?.Text;
Assert.That(t, Is.Not.Null);
Assert.That(t.Length, Is.GreaterThan(0));
```

I am a bit disappointed that the following also fails:
```
A? a = GetA(create);
Assert.That(a?.B, Is.Not.Null);
B b = a.B;
string? t = b.Text;
Assert.That(t, Is.Not.Null);
Assert.That(t.Length, Is.GreaterThan(0));
```

The assignment `B b = a.B` where `b` is not nullable is suppressed correctly, but then the compiler still complains about `b` being possible `null` in `b.Text`. It looks like suppressors can only suppress an error, but not alter the compilers idea of what could be `null`.
It works when doing:
```
A? a = GetA(create);
B? b = a?.B;
Assert.That(b, Is.Not.Null);
string? t = b.Text;
Assert.That(t, Is.Not.Null);
Assert.That(t.Length, Is.GreaterThan(0));
```

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs
Comment (manfred-brands) [reply src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:84]: Correct. I have removed it.

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DiagnosticsSuppressorAnalyzer.cs
Comment (manfred-brands) [reply src/nunit.analyzers.tests/DiagnosticSuppressors/DiagnosticsSuppressorAnalyzer.cs:106]: Removed.

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DiagnosticsSuppressorAnalyzer.cs
Comment (manfred-brands) [reply src/nunit.analyzers.tests/DiagnosticSuppressors/DiagnosticsSuppressorAnalyzer.cs:86]: All our target frameworks predate nullability. One way to get the correct annotations in is to use the excellent [TunnelVisionLabs.ReferenceAssemblyAnnotator](https://github.com/tunnelvisionlabs/ReferenceAssemblyAnnotator) to pull in .NET5.0 annotated references.
I did that and guess what, it complained about the `programmaticSuppressionInfoProperty` being null. So good catch by you. It also reported a few more.
I added two commits, one with updating the .csproj files and one with fixing the violations.

**Comment by manfred-brands** [comment]
> I'm most happy to merge this when you are

The one thing I didn't add was documentation. As it suppresses compiler errors, it doesn't raise an issue by itself.
But maybe someone wants to disable the suppressors as it makes their code fail when analyzers are off. On our big project we have Live analyzers enabled (while editing in VS)so that code that gets touched, gets improved, but they are disabled during builds as older code doesn't (yet) comply. This means code won't compile even though it doesn't show errors in VS.
Disabling the suppression is a hassle until [roslyn#4972](https://github.com/dotnet/roslyn/issues/4972) is resolved.
It requires a separate .ruleset 

I added two documentation files and updated the index.md
This caught that I should have started the new range with NUnit3001 to be consistent with the others: Fixed.

The documentation is not yet tested in DocumentationTests.

**Comment by manfred-brands** [comment]
> The documentation is not yet tested in DocumentationTests.

It is now.

**Review by mikkelbu — APPROVED** [review]
Again, great work. I've added some nitpick (mostly about the documentation), but besides this I think makes a great addition. As before I'm - even more - happy to merge this, when you are ready.

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs
Comment (manfred-brands) [reply src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:91]: This is needed to prevent a compile error (_warning CS0649: Field 'TestClass.s' is never assigned to, and will always have its default value null_) which causes the unit test to fail.

File: src/nunit.analyzers.tests/DocumentationTests.cs
Comment (manfred-brands) [reply src/nunit.analyzers.tests/DocumentationTests.cs:498]: Yes, added. And it wound an issue in the documentation using the wrong description in the .editorconfig section.

**Comment by manfred-brands** [comment]
Thanks @mikkelbu for the review. Ready for merge.

File: src/nunit.analyzers.tests/DocumentationTests.cs
Comment (mikkelbu) [reply src/nunit.analyzers.tests/DocumentationTests.cs:500]: Great 👍

File: src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs
Comment (mikkelbu) [reply src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldIsUninitializedSuppressor.cs:36]: Thanks for the explanation. It helped a lot on my understanding.

**Pull request merged**
